README.md
# homebridge-landroid [![NPM Version](https://img.shields.io/npm/v/homebridge-landroid.svg)](https://www.npmjs.com/package/homebridge-landroid) [![verified-by-homebridge](https://badgen.net/badge/homebridge/verified/purple)](https://github.com/homebridge/homebridge/wiki/Verified-Plugins)
Homebridge plugin to control Worx Landroid (as well as Kress, Ferrex and Landxcape) lawn mowers through the Cloud, should support most mowers.

## Features
 - Automatically fetches all mowers from Cloud
 - Start mower
 - Return mower to home
 - Mowing status (on / off)
 - Battery Status
 - Error status
 - Rain status
 - Home status

## Installation
1. Install homebridge using: npm install -g homebridge
2. Install this plugin using: npm install -g homebridge-landroid
3. Set up homebridges config.json with your Worx account data

#### Example config
```
{
  "bridge": {
    "name": "Homebridge",
    "username": "CC:22:3D:E3:CE:30",
    "port": 51826,
    "pin": "031-45-154"
  },
  "platforms": [
    {
      "platform": "Landroid",
      "email": "my@email.com",
      "pwd": "my_password"
    }
  ]
}
```

#### Options
 - `email` eMail for your Worx account
 - `pwd` Password for your Worx account
 - `rainsensor` Adds an additional "Leak" sensor for rain detection
 - `homesensor` Adds an additional "Home" sensor for home detection
 - `partyswitch` Adds an additional "PartyMode" switch to temporarily disable the schedule
 - `reload` Clears all mowers in HomeKit and reloads them from the cloud, default `false`
 - `cloud` Sets the cloud to use, `worx`, `kress`, `ferrex` or `landxcape`, default `worx`
 - `debug` Enable additional debug log output, default `false`
 - `mowdata` Enable additional mowing data log output, default `false`

## Usage
 The mower will appear as a switch and a contact sensor in HomeKit.

#### On/Off Switch
The switch shows the current status and allows to control the mower. If the switch is off the mower is either on the home base or on its way to the home base. If it's on the mower is currently mowing. Turn the switch on to start the mowing cycle, turn it off to send the mower back home.

#### Contact Sensor
The contact sensor is used to display issues with the mower (trapped, outside wire etc.), when the contact sensor is "open" there is some issue that prevents the mower from continuing. Fix the issue to control the mower again.

#### Home Sensor
The home (contact) sensor is used to display if the mower is standing in home position, when the contact sensor is "open" the mower is currently not at home.

Note: If you update from Version <= 0.9.5 or enable the sensor after first setup then you need to reload all mowers. This also invalids your homekit automations which need to be reconfigured.

#### Battery Status
You can see the battery status in the settings of either the switch or contact sensor in the Home app and you can ask Siri about the battery status of your lawn mower.

## Development
If you want new features or improve the plugin, you're very welcome to do so. The projects `devDependencies` include homebridge and the `npm run test` command has been adapted so that you can run a test instance of homebridge during development. Alternatively you can use `npm run dev` if you have homebridge installed as a global package
#### Setup
- clone github repo
- `npm install` in the project folder
- create `.homebridge` folder in project root
- add `config.json` with appropriate content to `.homebridge` folder
- run `npm run test` to start the homebridge instance for testing
- alternatively run `npm run dev` to start homebridge if its installed as a global package

#### Notes
Most of the connector code is directly copied from [iobroker.worx](https://github.com/iobroker-community-adapters/ioBroker.worx), a helper script called 'sync-code' allows updating to the latest version of that code.


CHANGELOG.md
# Changelog
This is the change log for the plugin, all relevant changes will be listed here.

For documentation please see the [README](https://github.com/normen/homebridge-landroid/blob/master/README.md)

## 0.12.4
- Remove unneccessary library import

## 0.12.3
- Avoid creating empty [object Object] mower

## 0.12.2
- Update worx endpoint and upstream library

## 0.11.9
- PartyMode switch now properly reflects PartyMode state

## 0.11.8
- Attempt to fix party mode status

## 0.11.7
- Another attempt at fixing error state

## 0.11.6
- Fix reading of error state
- Update from upstream

## 0.11.5
- Better handle multiple mowers (upstream)

## 0.11.4
- Improve adapter to avoid unexpected timer issues

## 0.11.3
- More logging, try to avoid issues with two mowers (upstream)

## 0.11.2
- Fix party mode status

## 0.11.1
- Persist cloud data

## 0.11.0
- Save session data

## 0.10.12
- Fix in value setting (upstream)

## 0.10.11
- Fix in cloud object storage

## 0.10.10
- Add request counter (upstream)

## 0.10.9
- Better compatibility to connector code (store cloud data)

## 0.10.8
- Change client id / username from iobroker to homebridge

## 0.10.7
- Fix removing old mowers

## 0.10.6
- Small fixes in login sequence
- More debug output
- Bump node dependency

## 0.10.5
- Small value reading fixes from upstream (iobroker.worx)

## 0.10.4
- Fix setTimeout error

## 0.10.3
- Fix missing value names in log output

## 0.10.2
- Fix status display

## 0.10.1
- Fix battery display

## 0.10.0
- New cloud connector

## 0.9.9
- Fix cloud login (no data yet)

## 0.9.8
- Update worx library to fix 404 Error

## 0.9.7
- Update worx library

## 0.9.6
- Added optional home sensor

## 0.9.5
- fix for (untested) support for "Party Mode"

## 0.9.4
- cleanup logging (thanks andy-dinger!)
- add (untested) support for "Party Mode"

## 0.9.3
- add cloud parameter to config panel

## 0.9.2
- fix https error
- possibly allow using kress and landxcape models

## 0.9.1
- downgrade/fix version of iobroker.worx library to 1.0.2

## 0.9.0    
- added mowing data option (thanks andy-dinger!)
- small cleanups and UI improvements
- add development infos
## 0.8.1

- fix name of rain sensor

## 0.8.0
- automatic removal of old mowers

## 0.7.3
- add optional leak sensor for rain detection

## 0.7.2
- fix reachable

## 0.7.1
- fix naming

## 0.7.0
- add mower auto-discovery

## 0.6.6
- always show cloud reported mower names by default

## 0.6.5
- add debug mode with additional log output

## 0.6.3
- fix import crash
- README cleanups

## 0.6.2
- fix sending messages
- clean up configuration panel

## 0.6.1
- use iobroker.worx library (Fixes #10)
- use global worx account (Fixes #11)

## 0.6.0
- use global email/pwd (per-mower settings will still work)

## 0.5.3
- support config-ui-x settings panel

## 0.5.2
- update iobroker.Landroid-s library (Fixes Issue #2)

## 0.5.1
- fix contact sensor state issue

## 0.5.0
- update README
- cleanups

## 0.4.3
- fix on/off state update
- searching home == off

## 0.4.2
- add logging for incoming status changes

## 0.4.1
- fix crash when login data is invalid

## 0.4.0
- fix sending commands
- move error status to contact sensor for now

## 0.2.0
- add update callbacks to HomeKit on changes
- fix battery service
- add charging state



CloudAdapter.js
/* adapter for iobroker plugin code */
const fs = require("fs");
const path_ad = require("path");
function Adapter(options) {
  this.objects = {};
  this.states = {};
  this.foreigns = {};
}

Adapter.prototype.loadLandroidObjectData = function(path) {
  this.instanceFile = path_ad.join(path, "session.json"); // TODO: not needed after move to object store
  this.object_file = path_ad.join(path,"objects.json");
  this.state_file = path_ad.join(path,"states.json");
  this.foreign_file = path_ad.join(path,"foreigns.json");
  if (fs.existsSync(this.object_file)) {
    this.objects = JSON.parse(fs.readFileSync(this.object_file, "utf8"));
  }
  if (fs.existsSync(this.state_file)) {
    this.states = JSON.parse(fs.readFileSync(this.state_file, "utf8"));
  }
  if (fs.existsSync(this.foreign_file)) {
    this.foreigns = JSON.parse(fs.readFileSync(this.foreign_file, "utf8"));
  }
  return this.states;
}

Adapter.prototype.saveLandroidObjectData = function() {
  // TODO: make sure to save on unexpected shutdown
  clearTimeout(this.saveTimeout);
  this.saveTimeout = setTimeout(this.saveLOD.bind(this), 3000);
}

Adapter.prototype.saveLOD = function() {
  fs.writeFileSync(this.object_file, JSON.stringify(this.objects));
  fs.writeFileSync(this.state_file, JSON.stringify(this.states));
  fs.writeFileSync(this.foreign_file, JSON.stringify(this.foreigns));
}

Adapter.prototype.on = function(o,v) { }
Adapter.prototype.extendObjectAsync = async function(o,v) { }
Adapter.prototype.setObjectNotExistsAsync = async function(o, v) { this.setObjectNotExists(o,v); };
Adapter.prototype.getObject = function(o) { if(this.objects[o]) {return this.objects[o];} else {return null;} }
Adapter.prototype.getObjectAsync = async function(o) { this.getObject(o);}
Adapter.prototype.setStateAsync = async function(o, v) { this.setState(o,v); };
Adapter.prototype.getState = function(o) { if(this.states[o]) {return this.states[o];} else {return null;} }
Adapter.prototype.getStateAsync = async function(o) { this.getState(o); }
Adapter.prototype.setForeignObject = function(o,v) { this.foreigns[o] = v; this.saveLandroidObjectData(); }
Adapter.prototype.delForeignObject = function(o) { delete this.foreigns[o]; }
Adapter.prototype.setForeignObjectAsync = async function(o,v) { this.setForeignObject(o,v); }
Adapter.prototype.delForeignObjectAsync = async function(o) { this.delForeignObject(o); }
Adapter.prototype.subscribeStates = function(o) { }
Adapter.prototype.setInterval = function(funct,delay){ return setInterval(funct.bind(this), delay); }
Adapter.prototype.setTimeout = function(funct,delay){ return setTimeout(funct.bind(this), delay); }
Adapter.prototype.clearInterval = function(obj){ clearInterval(obj); }
Adapter.prototype.clearTimeout = function(obj){ clearTimeout(obj); }

module.exports = Adapter;


LandroidDataset.js
"use strict";
const moment = require("moment");

function LandroidDataset(readings){
  if(readings != null && readings != undefined)
    this.parse(readings);
}

LandroidDataset.prototype.parse = function(readings) {
    if (readings["cfg"]) {
        this.language = readings["cfg"]["lg"];
        this.dateTime = moment(readings["cfg"]["dt"] + " " + readings["cfg"]["tm"], "DD/MM/YYYY HH:mm:ss");
        this.rainDelay = parseInt(readings["cfg"]["rd"], 10);
        this.serialNumber = readings["cfg"]["sn"];
        if (readings["cfg"]["sc"]) {
            this.active = (readings["cfg"]["sc"]["m"] ? true : false);
            this.partyMode = (readings["cfg"]["sc"]["m"] === 2 && typeof(readings["cfg"]["sc"]["distm"]) !== 'undefined');
            this.timeExtension = Number(readings["cfg"]["sc"]["p"]).valueOf();
            if (readings["cfg"]["sc"]["d"]) {
                this.schedule = [];
                let entries = readings["cfg"]["sc"]["d"];
                entries.forEach(entry => {
                    /*let timePeriod = new TimePeriod();
                    let start = String(entry[0]).split(":");
                    timePeriod.startHour = parseInt(start[0], 10);
                    timePeriod.startMinute = parseInt(start[1], 10);
                    timePeriod.durationMinutes = parseInt(entry[1], 10);
                    timePeriod.cutEdge = (entry[2] ? true : false);
                    this.schedule.push(timePeriod);*/
                });
            }
        }
    }
    if (readings["dat"]) {
        if (readings["dat"]["st"]) {
            this.totalTime = Number(readings["dat"]["st"]["wt"]).valueOf();
            this.totalDistance = Number(readings["dat"]["st"]["d"]).valueOf();
            this.totalBladeTime = Number(readings["dat"]["st"]["b"]).valueOf();
        }
        if (readings["dat"]["bt"]) {
            this.batteryChargeCycle = Number(readings["dat"]["bt"]["nr"]).valueOf();
            this.batteryCharging = (readings["dat"]["bt"]["c"] ? true : false);
            this.batteryVoltage = Number(readings["dat"]["bt"]["v"]).valueOf();
            this.batteryTemperature = Number(readings["dat"]["bt"]["t"]).valueOf();
            this.batteryLevel = Number(readings["dat"]["bt"]["p"]).valueOf();
        }
        this.macAddress = readings["dat"]["mac"];
        this.firmware = readings["dat"]["fw"];
        this.wifiQuality = Number(readings["dat"]["rsi"]).valueOf();
        this.statusCode = Number(readings["dat"]["ls"]).valueOf();
        this.statusDescription = LandroidDataset.STATUS_CODES[this.statusCode];
        this.errorCode = Number(readings["dat"]["le"]).valueOf();
        this.errorDescription = LandroidDataset.ERROR_CODES[this.errorCode];
    }
}

LandroidDataset.STATUS_CODES = {
    0: 'IDLE',
    1: 'Home',
    2: 'Start sequence',
    3: 'Leaving home',
    4: 'Follow wire',
    5: 'Searching home',
    6: 'Searching wire',
    7: 'Mowing',
    8: 'Lifted',
    9: 'Trapped',
    10: 'Blade blocked',
    11: 'Debug',
    12: 'Remote control',
    30: 'Going home',
    31: 'Zone training',
    32: 'Border Cut',
    33: 'Searching zone',
    34: 'Pause'
};

LandroidDataset.ERROR_CODES = {
    0: 'No error',
    1: 'Trapped',
    2: 'Lifted',
    3: 'Wire missing',
    4: 'Outside wire',
    5: 'Raining',
    6: 'Close door to mow',
    7: 'Close door to go home',
    8: 'Blade motor blocked',
    9: 'Wheel motor blocked',
    10: 'Trapped timeout',
    11: 'Upside down',
    12: 'Battery low',
    13: 'Reverse wire',
    14: 'Charge error',
    15: 'Timeout finding home',
    16: 'Mower locked',
    17: 'Battery temperature out of range',
};

module.exports = LandroidDataset;


index.js
"use strict";
/*
 * Landroid Worx plugin for HomeBridge
 *
 * CloudConnector and all files in the lib folder are copied from
 * https://github.com/iobroker-community-adapters/ioBroker.worx
 *
 */
const fs = require("fs");
const path_ad = require("path");
var Accessory, Service, Characteristic, UUIDGen, STORAGE_PATH;
var CloudConnector = require('./CloudConnector');
var LandroidDataset = require('./LandroidDataset');

function LandroidPlatform(log, config, api) {
  this.config = config;
  this.log = log;
  this.partymode = config.partymode || false;
  this.debug = config.debug || false;
  this.mowdata = config.mowdata || false;
  this.cloud = config.cloud || "worx";
  this.accessories = [];
  this.cloudMowers = [];

  if(!config.email || !config.pwd){
    this.log("WARNING: No account configured, please set email and password of your Worx account in config.json!");
    return;
  }

  const self = this;
  this.api = api;

  // Listen to event "didFinishLaunching", this means homebridge already finished loading cached accessories.
  // Platform Plugin should only register new accessory that doesn't exist in homebridge after this event.
  // Or start discover new accessories.
  this.api.on('didFinishLaunching', async function () {
    self.log('DidFinishLaunching');
    if(self.config.reload){
      self.log('**** WARNING: Landroid plugin is in reload mode, mowers will be recreated each boot ****');
      self.accessories.forEach(accessory => {
        self.log('Removing Landroid ' + accessory.accessory.displayName + ' from HomeKit');
        self.api.unregisterPlatformAccessories('homebridge-landroid', 'Landroid', [accessory.accessory]);
      });
      self.accessories = [];
    }
    self.landroidCloud = CloudConnector();
    let persisted_states = self.landroidCloud.loadLandroidObjectData(STORAGE_PATH);
    for(var myname in persisted_states) {
      self.createUpdate(myname, persisted_states[myname]);
    }

    self.accessories.forEach(accessory=>{
      accessory.landroidCloud = self.landroidCloud;
    });
    self.landroidCloud.config = {
      "mail": config.email,
      "password": config.pwd,
      "server": config.cloud || "worx"
    }
    self.landroidCloud.log = new LandroidLogger(log);
    self.landroidCloud.log.isDebug = self.debug;
    self.landroidCloud.setState = function(objectname,object) {
      self.landroidCloud.states[objectname] = object;
      self.landroidCloud.saveLandroidObjectData();
      if(objectname == "info.connection" && object == true){
        self.removeTimeout = setTimeout(self.clearOldMowers.bind(self), 60000);
      } else if(objectname.includes(".mower.")){
        let serial = objectname.substring(0, objectname.indexOf("."));
        let item = objectname.split('.').pop();

        if(object && (object.val !== null || object.val !== undefined)) {
          self.landroidUpdate(serial, item, object.val);
        }
      }
    };
    self.landroidCloud.setObjectNotExists = async function(objectname, object) {
      if(!objectname instanceof String) return;
      if(!object.common.name instanceof String) return;
      if(!self.landroidCloud.objects[objectname]) {
        self.landroidCloud.objects[objectname] = {};
      }
      if(!objectname.includes(".") && object.common.name != "[object Object]"){
        self.landroidFound(object.common.name, objectname);
      }
    };
    await self.landroidCloud.onReady();
  });
}

LandroidPlatform.prototype.clearOldMowers = function() {
  const self = this;
  this.accessories.forEach((accessory, idx, obj) => {
    if(!self.cloudMowers.includes(accessory.serial)){
      self.api.unregisterPlatformAccessories('homebridge-landroid', 'Landroid', [accessory.accessory]);
      obj.splice(idx, 1);
    }
  });
}

// Function invoked when homebridge tries to restore cached accessory.
LandroidPlatform.prototype.configureAccessory = function(accessory) {
  if (!this.config) { // happens if plugin is disabled and still active accessories
    return;
  }
  this.log('Restoring Landroid ' + accessory.displayName + ' from HomeKit');
  accessory.reachable = false;
  this.accessories.push(new LandroidAccessory(this, null, null, accessory));
}

// Handler will be invoked when user try to config your plugin.
// Callback can be cached and invoke when necessary.
LandroidPlatform.prototype.configurationRequestHandler = function(context, request, callback) {
  callback(null);
}

LandroidPlatform.prototype.landroidFound = function(name, serial) {
  if(this.debug) {
    this.log("[DEBUG] MOWER: " + name + " (" + serial + ")");
  }else {
    this.log("Found Landroid in Worx Cloud with name: " + name);
  }
  if(this.mowdata) {
    this.log("Mowing data logging enabled for Landroid " + name);
  }
  if(!this.cloudMowers.includes(serial)){
    this.cloudMowers.push(serial);
  }
  for(var i = 0; i<this.accessories.length; i++){
    const accessory = this.accessories[i];
    if(accessory.serial == serial){
      //already have this one
      accessory.accessory.reachable = true;
      //this.landroidUpdate(mower, data);
      return;
    }
  }
  // don't have this one, add it
  const newMower = new LandroidAccessory(this, name, serial);
  this.accessories.push(newMower);
  this.log("Adding Landroid " + name + " to HomeKit");
  this.api.registerPlatformAccessories('homebridge-landroid', 'Landroid', [newMower.accessory]);
  //this.landroidUpdate(mower,data);
}

LandroidPlatform.prototype.createUpdate = function(objectname, object) {
  if(objectname.includes(".mower.")){
    let serial = objectname.substring(0, objectname.indexOf("."));
    let item = objectname.split('.').pop();
    if(object && (object.val !== null || object.val !== undefined)) {
      this.landroidUpdate(serial, item, object.val);
    }
  }
}

LandroidPlatform.prototype.landroidUpdate = function(serial, item, data) {
    if(this.debug) {
      this.log("[DEBUG] DATA: " + item + ": " + JSON.stringify(data));
    }
    this.accessories.forEach(accessory=>{
        accessory.landroidUpdate(serial, item, data, this.mowdata);
    });
}

//TODO: add other constructor
function LandroidAccessory(platform, name, serial, accessory) {
    this.landroidCloud = platform.landroidCloud;
    this.log = platform.log;
    this.config = platform.config;

    if (accessory) {
      this.accessory = accessory;
    } else {
      // new accessory object
      var uuid = UUIDGen.generate(serial);
      //this.log('Creating new accessory for ' + name + ' (' + serial + ')');
      this.accessory = new Accessory("Landroid " + name, uuid);
      this.accessory.context.name = name;
      this.accessory.context.serial = serial;

      this.accessory.addService(new Service.Switch("Landroid " + name));
      this.accessory.addService(new Service.BatteryService());
      this.accessory.addService(new Service.ContactSensor("Landroid " + name + " Problem", "ErrorSensor"));
      if(this.config.rainsensor) this.accessory.addService(new Service.LeakSensor("Landroid " + name + " Rain"));
      if(this.config.homesensor) this.accessory.addService(new Service.ContactSensor("Landroid " + name + " Home", "HomeSensor"));
      if(this.config.partymode) this.accessory.addService(new Service.Switch("Landroid " + name + " PartyMode", "PartySwitch"));
    }

    this.name = this.accessory.context.name;
    this.serial = this.accessory.context.serial;

    this.dataset = {};
    this.dataset.batteryState = 0;
    this.dataset.batteryCharging = false;
    this.dataset.statusCode = 0;
    this.dataset.errorCode = 0;

    if(this.accessory.getService("Landroid " + name)){
      this.accessory.getService("Landroid " + name).getCharacteristic(Characteristic.On).on('get', this.getOn.bind(this));
      this.accessory.getService("Landroid " + name).getCharacteristic(Characteristic.On).on('set', this.setOn.bind(this));
    } else{
      this.log("Fallback for On/Off switch");
      this.accessory.getService(Service.Switch).getCharacteristic(Characteristic.On).on('get', this.getOn.bind(this));
      this.accessory.getService(Service.Switch).getCharacteristic(Characteristic.On).on('set', this.setOn.bind(this));
    }

    this.accessory.getService(Service.BatteryService).getCharacteristic(Characteristic.BatteryLevel).on('get', this.getBatteryLevel.bind(this));
    this.accessory.getService(Service.BatteryService).getCharacteristic(Characteristic.StatusLowBattery).on('get', this.getStatusLowBattery.bind(this));
    this.accessory.getService(Service.BatteryService).getCharacteristic(Characteristic.ChargingState).on('get', this.getChargingState.bind(this));

    if(this.accessory.getService("ErrorSensor")){
      this.accessory.getService("ErrorSensor").getCharacteristic(Characteristic.ContactSensorState).on('get', this.getContactSensorStateError.bind(this));
    } else{
      this.accessory.getService(Service.ContactSensor).getCharacteristic(Characteristic.ContactSensorState).on('get', this.getContactSensorStateError.bind(this));
    }
  
    this.accessory.getService(Service.AccessoryInformation)
      .setCharacteristic(Characteristic.Name, this.name)
      .setCharacteristic(Characteristic.Manufacturer, 'Worx')
      .setCharacteristic(Characteristic.Model, 'Landroid')
      .setCharacteristic(Characteristic.SerialNumber, this.serial);

    if(this.config.rainsensor && this.accessory.getService(Service.LeakSensor)) this.accessory.getService(Service.LeakSensor).on('get', this.getLeak.bind(this));
    if(this.config.homesensor && this.accessory.getService("HomeSensor")) this.accessory.getService("HomeSensor").getCharacteristic(Characteristic.ContactSensorState).on('get', this.getContactSensorStateHome.bind(this));
    if(this.config.partymode && this.accessory.getService("PartySwitch")) {
      this.accessory.getService("PartySwitch").getCharacteristic(Characteristic.On).on('get', this.getPartyMode.bind(this));
      this.accessory.getService("PartySwitch").getCharacteristic(Characteristic.On).on('set', this.setPartyMode.bind(this));
    } else if(this.config.partymode) {
      this.log("Party switch not found");
    }
}

LandroidAccessory.prototype.landroidUpdate = function(serial, item, data, mowdata) {
  var totalTime, totalBladeTime, totalDistance;

  if(serial !== this.serial) return;

  if(data != null && data != undefined){
    let oldDataset = this.dataset;
    if(item == "status") {
      item = "statusCode";
    } else if(item == "error") {
      item = "errorCode";
    }
    this.dataset[item] = data;
    // this.log("landroidUpdate ran with RSSI " + this.dataset.wifiQuality + ", battery temperature " + this.dataset.batteryTemperature);
    if(mowdata){
      if(oldDataset.totalTime == null || oldDataset.totalTime == undefined){
         // Initialise mowing data
        this.saveTime = Number(this.dataset.totalTime);
        totalTime = this.saveTime / 60;
        totalTime = totalTime.toFixed(2);
        this.saveBladeTime = Number(this.dataset.totalBladeTime);
        totalBladeTime = this.saveBladeTime / 60;
        totalBladeTime = totalBladeTime.toFixed(2);
        this.saveDistance = Number(this.dataset.totalDistance);
        this.log("Landroid " + this.name + " hours worked so far: " + totalTime 
          + ", hours mowed so far: " + totalBladeTime 
          + ", distance moved so far: " + String(this.saveDistance / 1000) + "km");
      }
    }
    if(this.dataset.batteryState != oldDataset.batteryState){
    //  this.log("Landroid " + this.name + " battery level changed to " + this.dataset.batteryState);
      this.accessory.getService(Service.BatteryService).getCharacteristic(Characteristic.BatteryLevel).updateValue(this.dataset.batteryState);
    }
    if(this.dataset.partyModus != oldDataset.partyModus){
      if(this.accessory.getService("PartySwitch")){
       this.accessory.getService("PartySwitch").getCharacteristic(Characteristic.On).updateValue(this.dataset.partyModus == true);
      }
    }
    if(this.dataset.batteryCharging != oldDataset.batteryCharging){
      this.log("Landroid " + this.name + " charging status changed to " + this.dataset.batteryCharging 
        + ", battery level " + this.dataset.batteryState);
      this.accessory.getService(Service.BatteryService).getCharacteristic(Characteristic.ChargingState).updateValue(this.dataset.batteryCharging?
        Characteristic.ChargingState.CHARGING:Characteristic.ChargingState.NOT_CHARGING);
    }
    if(this.dataset.statusCode != oldDataset.statusCode){
      this.log("Landroid " + this.name + " status changed to " + this.dataset.statusCode + " (" + this.dataset.statusDescription + ")" 
        + ", battery level " + this.dataset.batteryState);
      if(isOn(this.dataset.statusCode)){
        this.accessory.getService(Service.Switch).getCharacteristic(Characteristic.On).updateValue(true);
        if(this.config.homesensor && this.accessory.getService("HomeSensor")) this.accessory.getService("HomeSensor").getCharacteristic(Characteristic.ContactSensorState).updateValue(Characteristic.ContactSensorState.CONTACT_NOT_DETECTED);
      }else{
        this.accessory.getService(Service.Switch).getCharacteristic(Characteristic.On).updateValue(false);
      }
      if(this.dataset.statusCode == 1 && oldDataset.totalTime != null && oldDataset.totalTime != undefined && mowdata) {
        // Landroid has just arrived home so show how much it's worked since last leaving (or last restarting Homebridge)
        totalTime = Number(this.dataset.totalTime);
        totalBladeTime = Number(this.dataset.totalBladeTime);
        totalDistance = Number(this.dataset.totalDistance);
        this.log("Landroid " + this.name + " new minutes worked: " + String(totalTime - this.saveTime)
          + ", new minutes mowed: " + String(totalBladeTime - this.saveBladeTime)
          + ", new distance moved: " + String(totalDistance - this.saveDistance) + "m");
        this.saveTime = totalTime;
        this.saveBladeTime = totalBladeTime;
        this.saveDistance = totalDistance;
      }
      if(this.dataset.statusCode == 1 && this.config.homesensor && this.accessory.getService("HomeSensor")) this.accessory.getService("HomeSensor").getCharacteristic(Characteristic.ContactSensorState).updateValue(Characteristic.ContactSensorState.CONTACT_DETECTED);
    }
    if(this.dataset.errorCode != oldDataset.errorCode){
      this.log("Landroid " + this.name + " error code changed to " + this.dataset.errorCode + " (" + this.dataset.errorDescription + ")" 
        + ", battery level " + this.dataset.batteryState);
      if(this.accessory.getService("ErrorSensor")){
        this.accessory.getService("ErrorSensor").getCharacteristic(Characteristic.ContactSensorState).updateValue(isError(this.dataset.errorCode)?
        Characteristic.ContactSensorState.CONTACT_NOT_DETECTED:Characteristic.ContactSensorState.CONTACT_DETECTED);
      } else{
        this.accessory.getService(Service.ContactSensor).getCharacteristic(Characteristic.ContactSensorState).updateValue(isError(this.dataset.errorCode)?
        Characteristic.ContactSensorState.CONTACT_NOT_DETECTED:Characteristic.ContactSensorState.CONTACT_DETECTED);
      }    
      if(this.config.rainsensor && this.accessory.getService(Service.LeakSensor)) this.accessory.getService(Service.LeakSensor).getCharacteristic(Characteristic.LeakDetected).updateValue(this.dataset.errorCode == 5);
    }
  }
}

LandroidAccessory.prototype.getContactSensorStateError = function(callback) {
  callback(null,  isError(this.dataset.errorCode)?Characteristic.ContactSensorState.CONTACT_NOT_DETECTED:Characteristic.ContactSensorState.CONTACT_DETECTED);
}

LandroidAccessory.prototype.getContactSensorStateHome = function(callback) {
  callback(null,  this.dataset.statusCode == 1?Characteristic.ContactSensorState.CONTACT_DETECTED:Characteristic.ContactSensorState.CONTACT_NOT_DETECTED);
}

LandroidAccessory.prototype.getBatteryLevel = function(callback) {
  callback(null, this.dataset.batteryState);
}

LandroidAccessory.prototype.getChargingState = function(callback) {
  callback(null, this.dataset.batteryCharging?Characteristic.ChargingState.CHARGING:Characteristic.ChargingState.NOT_CHARGING);
}

LandroidAccessory.prototype.getStatusLowBattery = function(callback) {
  callback(null, this.dataset.errorCode == 12?Characteristic.StatusLowBattery.BATTERY_LEVEL_LOW:Characteristic.StatusLowBattery.BATTERY_LEVEL_NORMAL);
}

LandroidAccessory.prototype.getOn = function(callback) {
  if(isOn(this.dataset.statusCode)){
    callback(null, true);
  }else{
    callback(null, false);
  }
}

LandroidAccessory.prototype.getLeak = function(callback) {
  if(this.dataset.statusCode == 5){
    callback(null, true);
  }else{
    callback(null, false);
  }
}

LandroidAccessory.prototype.setOn = function(state, callback) {
  if(state){
    this.sendMessage(1);
  }else{
    this.sendMessage(3);
  }
  callback(null);
}

LandroidAccessory.prototype.getPartyMode = function(callback) {
  if(this.dataset.partyModus) {
    callback(null, true);
  }else{
    callback(null, false);
  }
}

LandroidAccessory.prototype.setPartyMode = function(state, callback) {
  if(!this.serial){
    this.log("Error: Mower has not been configured yet.");
  }
  let outMsg = "";
  if (state) {
    outMsg = '{"sc":{ "m":2, "distm": 0}}';
  } else{
    outMsg = '{"sc":{ "m":1, "distm": 0}}';
  }
  this.log("Sending to Landroid " + this.name + ": [" + outMsg + "] ("+this.serial+")");
  this.landroidCloud.sendMessage(outMsg, this.serial);
}

LandroidAccessory.prototype.sendMessage = function(cmd, params) {
  if(!this.serial){
    this.log("Error: Mower has not been configured yet.");
  }
  let message = {};
    if (cmd) {
        message["cmd"] = cmd;
    }
    if (params) {
        message = Object.assign(message, params);
    }
    let outMsg = JSON.stringify(message);
    this.log("Sending to Landroid " + this.name + ": [" + outMsg + "] ("+this.serial+")");
    this.landroidCloud.sendMessage(outMsg, this.serial);
}

function isOn(c){
  if(c == 2 || c == 3 || c == 4 || c == 6 || c == 7 || c == 32 || c == 33){
    return true;
  }else{
    return false;
  }
}

function isError(c){
  //no error and rain delay is "not an error"
  if(c == 0 || c == 5){
    return false;
  }else{
    return true;
  }
}

function LandroidLogger(log){
  let that = this;
  this.log = log;
  this.logMsg = function(msg){
    that.log(msg);
  }
  this.noLogMsg = function(msg){
    if(this.isDebug){
      that.log(msg);
    }
  }
  this.trace = this.noLogMsg;
  this.debug = this.noLogMsg;
  this.info = this.logMsg;
  this.warn = this.logMsg;
  this.error = this.logMsg;
  this.fatal = this.logMsg;
}

function updateStorage(newPath){
  var confPath = newPath + "/plugin-persist/homebridge-landroid";
  if(!fs.existsSync(confPath)){
    fs.mkdirSync(confPath, {recursive: true});
  }
  return confPath;
}

module.exports = function(homebridge) {
  Accessory = homebridge.platformAccessory;
  Service = homebridge.hap.Service;
  Characteristic = homebridge.hap.Characteristic;
  UUIDGen = homebridge.hap.uuid;
  STORAGE_PATH = updateStorage(homebridge.user.storagePath());
  homebridge.registerPlatform("homebridge-landroid", "Landroid", LandroidPlatform, true);
}
